text/0000-pointer-match.md

```diff
@@ -0,0 +1,282 @@
+- Feature Name: `pointer-match`
+- Start Date: 2019-03-21
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Extend match syntax and patterns by support for a limited set of operations for
+pointers, which involve only address calculation and not actually reading
+through the pointer value. Make it possible to use these matches to calculate
+addresses of fields even for `repr(packed)` structs and possibly unaligned
+pointers where an intermediate reference must not be created.
+
+# Motivation
+[motivation]: #motivation
+
+To create a pointer to a field of a struct, there is currently no way in Rust
+that avoids creating a temporary reference. Since reference semantics are
+stricter, this may lead to silent undefined behaviour where that reference
+should not be valid. Depending on the resolution of reference semantics this
+affects:
+
+* Creating a pointer to a field of a packed struct, where the reference may be
+  unaligned (depending on .
+* Pointing to fields of an uninitialized type, where the reference points to
+  uninitialized data. This may be complicated by unions, where it could be
+  possible that not a single variant is currently completely initialized, yet
+  one wants to access some subfield. See
+  <https://github.com/rust-lang/unsafe-code-guidelines/issues/73#issuecomment-460634637>.
+* Doing pointer offset calculations where the references does not refer to the
+  same, or any, allocation. This is because reference calculations are
+  performed with `getelementptr inbounds`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Match expression are extended from support for a reference binding mode, to a
+pointer binding mode. Furthermore, a new pattern binds to a pointer, and
+identifiers are extended to allow a new mode similar to `ref` and `ref mut`
+binding to a reference. These patterns are called pointer pattern and raw
+identifier for the remainder of the document.
+
+```
+#[repr(packed)]
+struct Foo {
+    a: u16,
+    b: u32,
+}
+
+fn ptr_b(foo: &mut Foo) -> *mut u32 {
+    let Foo { raw mut b, .. } = foo;
+    b
+}
+```
+
+Note that pointer binding mode and pointer pattern requires `unsafe`, even when
```

Comment:
There have been multiple references to these unsafe pointer patterns, but no examples or further explanations. What are they and why are they necessary? Aren't `raw mut` or `raw const` patterns sufficient?

Reply:
Pointer patterns are the counter part of reference patterns, necessary for disambiguation in some special cases. They use `* <subpattern>`, paralleling `& <subpattern>` and pointer binding mode automatically adds the top-level pointer pattern, the same as the reference pattern implied by reference binding mode. I'll add an example showing the necessity but it boils down to being able to match a pointer by-value and its content with `raw` pattern, the former being necessary for backwards compatibility.

Reply:
@oli-obk New section at the beginning of https://github.com/HeroicKatora/rfcs/blob/pointer-match/text/0000-pointer-match.md#reference-level-explanation to explain all of this